run/aem2john.py

```diff
@@ -1,74 +1,118 @@
+input, the script logs a warning and skips to the next hash. 
+
+Usage:
+    python aem2john.py <file1> [<file2> ...]
+
+Each argument should be a file path pointing to a text file containing AEM hashes, one per line.
+
+The script is compatible with both Python 2 and Python 3.
+
+Originally authored by Dhiru Kholia <kholia at kth.se> in 2018 and released to the general public under a permissive
+license allowing free redistribution and modification.
+
+For more details about the original AEM hash generation algorithm, see the "generateHash" method in PasswordUtil.java
+from the Apache Jackrabbit Oak project: https://github.com/apache/jackrabbit-oak.
+"""
+
+# Use the print function for Python 2 and 3 compatibility
+from __future__ import print_function
 
 import sys
+import logging
+from argparse import ArgumentParser
+
+# Map algorithm tags to their respective algorithm numbers
+ALGO_TAGS = {
+	"{SHA-256}": 3,  # SHA-256
+	"{SHA-512}": 4  # SHA-512
+}
+
+logging.basicConfig(level=logging.INFO, format='%(levelname)s: %(message)s')
+
+# Python 2 and 3 compatibility for file paths
+if sys.version_info.major == 2:
+	Path = str
+else:
+	from pathlib import Path
+
+
+def split_user_and_hash(line):
+	"""Split username and hash from the line."""
+	if ':' in line:
+		return line.split(':', 1)
+	return '', line
+
 
-PY3 = sys.version_info[0] == 3
+def extract_algo_and_hash(line):
+	"""Extract algorithm number and hash from the line."""
+	for tag, algo in ALGO_TAGS.items():
+		if tag in line:
+			return algo, line[len(tag):]
+	return None, line
 
-if not PY3:
-    reload(sys)
-    sys.setdefaultencoding('utf8')
 
-# Input,
-#   {SHA-256}a9d4b340cb43807b-1000-33b8875ff3f9619e6ae984add262fb6b6f043e8ff9b065f4fb0863021aada275
-#   jsmith:{SHA-256}fe90d85cdcd7e79c-1000-ef182cdc47e60b472784e42a6e167d26242648c6b2e063dfd9e27eec9aa38912
-#   admin:{SHA-512}fe90d85cdcd7e79c-1000-4c29a0ac964e7bbc5380797f294d15928288cbcde3d501eb8746296de8d6c06b2b5ff27b56ae174744fe69ee157614ad126c1315ee3b67c891e42753e01a3e37
-#
-# Output,
-#   $sspr$3$1000$a9d4b340cb43807b$33b8875ff3f9619e6ae984add262fb6b6f043e8ff9b065f4fb0863021aada275
-#   jsmith:$sspr$3$1000$fe90d85cdcd7e79c$ef182cdc47e60b472784e42a6e167d26242648c6b2e063dfd9e27eec9aa38912
-#   admin:$sspr$4$1000$fe90d85cdcd7e79c$4c29a0ac964e7bbc5380797f294d15928288cbcde3d501eb8746296de8d6c06b2b5ff27b56ae174744fe69ee157614ad126c1315ee3b67c891e42753e01a3e37
-#
-# Passwords -> admin, Aa12345678!@
+def split_hash_parts(line):
+	"""Split hash into salt, iterations, and actual hash."""
+	parts = line.split('-')
+	if len(parts) != 3:
+		return None
+	return tuple(parts)
+
+
+def process_line(line):
+	"""Process a single line from the input file."""
+	line = line.rstrip()
+	user, hash_with_algo = split_user_and_hash(line)
+	algo, hash_without_algo = extract_algo_and_hash(hash_with_algo)
+
+	if algo is None:
+		logging.warning(f"Unknown hash format in line: {line[:8]}")
```

Comment:
I love f-strings too, but the script claims to be python2 compatible and f-strings are python3.6+

Reply:
Oh, I guess this implies this revised script wasn't tested on Python below 3.6 at all, and the Python 2 compatibility is purely theoretical? I think we shouldn't be merging these PRs without actual testing across a range of Python versions.